Add Ability to Override PotionEffect Default Sort Order

Flashfyre opened the pull request:
Added getSortOrder to Potion class for overriding default potion effect sorting

Flashfyre committed:
Add Ability to Override PotionEffect Default Sort Order

Added getSortOrder to Potion class for overriding default potion effect
sorting

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/MinecraftForge/MinecraftForge?pullRequest=4172) <br/>All committers have signed the CLA.

Flashfyre commented:
I added this because my (very colourful) mod has potions with a liquid colour that changes over time. This causes my custom ambient potion effects to frequently shift order in the inventory GUI due to potion effects being sorted by liquid colour, which is very annoying and even distracting since potion effects can also be seen in the upper right.

williewillus commented on patches/minecraft/net/minecraft/potion/PotionEffect.java.patch:
```diff
@@ -41,6 +41,15 @@
+     {
+         int i = 32147;
+-        return (this.func_76459_b() <= 32147 || p_compareTo_1_.func_76459_b() <= 32147) && (!this.func_82720_e() || !p_compareTo_1_.func_82720_e()) ? ComparisonChain.start().compare(Boolean.valueOf(this.func_82720_e()), Boolean.valueOf(p_compareTo_1_.func_82720_e())).compare(this.func_76459_b(), p_compareTo_1_.func_76459_b()).compare(this.func_188419_a().func_76401_j(), p_compareTo_1_.func_18841 ...
++        return (this.func_188419_a().getSortIndex() == p_compareTo_1_.func_188419_a().getSortIndex()) ? (this.func_76459_b() <= 32147 || p_compareTo_1_.func_76459_b() <= 32147) && (!this.func_82720_e() || !p_compareTo_1_.func_82720_e()) ? ComparisonChain.start().compare(Boolean.valueOf(this.func_82720_e()), Boolean.valueOf(p_compareTo_1_.func_82720_e())).compare(this.func_76459_b(), p_compareTo_ ...
```
i think you could simplify this to the following:
```Java
+ int cmp = Integer.compare(this.getPotion().getSortIndex(), that.getPotion().getSortIndex());
+ if (cmp != 0) return cmp;
< leave vanilla logic here >
```

then you won't need to do a complicated patch on the vanilla logic

Flashfyre replied:
Will do.

Flashfyre committed:
Changed PotionEffect logic to avoid altering Vanilla logic

Changed PotionEffect logic to avoid altering Vanilla logic according ot
a recommendation from williewillus

liach commented on patches/minecraft/net/minecraft/potion/Potion.java.patch:
```diff
@@ -88,6 +88,18 @@
++     * 
++     * @return An index to override the default sort order
++     */
++    @SideOnly(Side.CLIENT)
```
Why? If it is comparable on the server then leave this.

Flashfyre replied:
I just didn't see a need for it anywhere on the server. I guess it wouldn't hurt to make it server-accessible.

liach commented on patches/minecraft/net/minecraft/potion/Potion.java.patch:
```diff
@@ -88,6 +88,18 @@
++     * @return An index to override the default sort order
++     */
++    @SideOnly(Side.CLIENT)
++    public int getSortIndex()
```
Can you make it potion effect aware?

Flashfyre replied:
Sure, good idea.

Flashfyre committed:
Made Potion.getSortIndex server-accessible and added a PotionEffect parameter

bs2609 commented on patches/minecraft/net/minecraft/potion/Potion.java.patch:
```diff
@@ -88,6 +88,18 @@
++     */
++    public int getSortIndex(PotionEffect potionEffect)
++    {
++    	return 0;
```
There's a tab here that needs replacing with spaces

Flashfyre replied:
My bad. Will fix.

Flashfyre committed:
Replaced tab in getSortIndex with spaces

Replaced tab in getSortIndex with spaces (tab was added by force of
habit)

howtonotwin commented on patches/minecraft/net/minecraft/potion/Potion.java.patch:
```diff
@@ -88,6 +88,18 @@
 +    }
++    
++    /**
++     * Return an index that will override the potion's default sort order. This is useful for potions with colors 
```
This isn't very clear if you don't look at the usage. Maybe add "The comparison between indices takes precedence over the vanilla logic. If two effects compare equal by index, vanilla takes over." after the first sentence.

Flashfyre replied:
Sure, I'll add some clarification.

howtonotwin commented on patches/minecraft/net/minecraft/potion/Potion.java.patch:
```diff
@@ -88,6 +88,18 @@
++     * @param The PotionEffect instance containing the potion
++     * @return An index to override the default sort order
++     */
++    public int getSortIndex(PotionEffect potionEffect)
```
"index" doesn't feel like the right word to use here. Maybe just "sortValue" or even "sortOverrideValue." Or maybe "index" really *is* the best choice.

Flashfyre replied:
I think index works fine, I've seen it used in many similar cases.

Flashfyre committed:
Added more detail to Potion.getSortIndex comments

mezz commented on patches/minecraft/net/minecraft/potion/Potion.java.patch:
```diff
@@ -88,6 +88,20 @@
++     * An index that, if different than that of a compared potion effect's potion, will override the default sort order.
++     * If the values are the same, the Vanilla logic alone will be responsible for sort order. Overriding this method is 
++     * useful for potions with changing liquid colors since, without this, the potion effect order would shift in the 
++     * inventory GUI when stacked with others due to Vanilla using liquid color as part of PotionEffect's compareTo logic.
```
This description is pretty long, I think it can be cut down to more basic information.
You should also mention some more functional information, such as the default value is 0 and whether higher indexes go earlier or later in the sort order.

Flashfyre replied:
Cool, I'll add this stuff tonight after work.

mezz commented on patches/minecraft/net/minecraft/potion/Potion.java.patch:
```diff
@@ -88,6 +88,20 @@
++     * @param The PotionEffect instance containing the potion
++     * @return An index to use to sort potion effects if the value is different than that of the other effect's potion
++     */
++    public int getSortIndex(PotionEffect potionEffect)
```
This name should probably have a clue about what is being sorted and where. `getSortIndexForGui` maybe?

Flashfyre replied:
Alright, I made the changes. Personally, I would prefer to be able to only override the liquid colour part, but I was told not to alter the Vanilla logic so I can make due with a total override.

mezz reviewed (changes requested):
Looks good, I think the documentation can be improved a little though.

Flashfyre committed:
Changed Potion.getSortIndex to getGuiSortIndex for clarity; changed description to better describe functionality

LexManos commented:
This screws up the entire sorting of lowest duration first.
Its a larger patch but this is better to be done as a replacement for the liquid color part specifically so you preserve the prioritization of remaining length and 'isAmbient'

Flashfyre commented:
I was originally going to do that but I was warned about altering vanilla logic. I will change it to override the liquid colour sorting.

Flashfyre committed:
Changed logic to only override the liquid color component of the PotionEffect compareTo

Flashfyre committed:
Changed logic to override liquid color as a sort parameter instead of adding an additional sort property

Flashfyre committed:
Remove last method comment line since the previous line already implies it

howtonotwin commented on patches/minecraft/net/minecraft/potion/Potion.java.patch:
```diff
@@ -88,6 +88,17 @@
++     * Used as a last resort for determining potion effect sort order in the gui.
++     * This can be used to preserve a potion effect's sort order if it has a liquid color that is subject to change.
++     * @param The PotionEffect instance containing the potion
++     * @return An index to use to sort potion effects if the value is different than that of the other effect's potion
```
This line is a bit clunky. Maybe just "a value used to sort `PotionEffect`s in GUIs. Defaults to this `Potion`'s color." Also, the same rules for capitalization and punctuation as used for `@param` are used for `@return`. [Reference](http://www.oracle.com/technetwork/java/javase/documentation/index-137868.html#@return)

howtonotwin commented on patches/minecraft/net/minecraft/potion/Potion.java.patch:
```diff
@@ -88,6 +88,17 @@
++    /**
++     * Used as a last resort for determining potion effect sort order in the gui.
++     * This can be used to preserve a potion effect's sort order if it has a liquid color that is subject to change.
++     * @param The PotionEffect instance containing the potion
```
This is invalid Javadoc. It needs to include the parameter name.  Also, it's convention to start this line with a lowercase letter if the first phrase contains no top-level verb. (That is, `@param something a blah`, not `@param something A blah`, and `@param something a blah that foos. Used to qux.`, not `@param something A blah that foos. Used to qux`, but use `@param something Does foo. Used to qux.`, not `@param something does foo. Used to qux`). [Reference](http://www.oracle.com/technetwork/java/javase/documentation/index-137868.html#@param)

Flashfyre replied:
Whoops, must've removed it when I was rewriting

howtonotwin commented on patches/minecraft/net/minecraft/potion/Potion.java.patch:
```diff
@@ -88,6 +88,17 @@
++    
++    /**
++     * Used as a last resort for determining potion effect sort order in the gui.
++     * This can be used to preserve a potion effect's sort order if it has a liquid color that is subject to change.
```
The Javadoc is inconsistent with "potion effect" and "PotionEffect". Please stick to one, and if you choose `PotionEffect`, enclose it in `{@code}` like `{@code PotionEffect}`.

howtonotwin reviewed (changes requested):
This is just nitpicking on the Javadoc. Also, for future reference, git commits should be written in the imperative mood (aka commands; like "Change logic to override color"), and the first line should be no longer than 50 chars.

Flashfyre committed:
Fix getGuiSortColor Javadoc

mezz commented on patches/minecraft/net/minecraft/potion/Potion.java.patch:
```diff
@@ -88,6 +88,17 @@
 +    }
++    
++    /**
++     * Used as a last resort for determining {@code PotionEffect} sort order in GUIs.
```
No need for the intensity here, it can just be
`Used for determining {@code PotionEffect} sort order in GUIs.`

mezz commented on patches/minecraft/net/minecraft/potion/Potion.java.patch:
```diff
@@ -88,6 +88,17 @@
++     */
++    public int getGuiSortColor(PotionEffect potionEffect)
++    {
++        return this.field_76414_N;
```
This should be `getLiquidColor()` to match the original logic (in case someone has overridden it for whatever reason?).

Flashfyre replied:
Good point.

mezz reviewed (changes requested):
Looks very clean now, just needs some minor fixes.

Flashfyre committed:
Change default return value to getter

LexManos commented:
Could be cleaner if you de-iifed it but sure.
